### PR TITLE
Class and function documentation for goto-symex-target [DOC-79]

### PR DESCRIPTION
**Pull request by xbauch**
Adding class and function documentation to goto-symex/symex_target. {h,cpp} and goto-symex/symex_target_equation.{h,cpp}.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/goto-symex/symex_target.cpp
Comment (majakusber) [inline src/goto-symex/symex_target.cpp]: According to the new coding standard, this should now live in the header:
```
For functions or methods that provide an interface to a module, a comment block should immediately precede the declaration of the entity it documents. This implies that it will live in the header file. 
```
:green_apple:

Comment (majakusber) [reply src/goto-symex/symex_target.cpp]: The word `generic` might be confusing (as in other parts of the codebase it refers to the Java generics). Consider using something like `A simple comparison..`

File: src/goto-symex/symex_assign.cpp
Comment (peterschrammel) [inline src/goto-symex/symex_assign.cpp]: I personally find `what` easier to read.

File: src/goto-symex/symex_assign.cpp
Comment (peterschrammel) [inline src/goto-symex/symex_assign.cpp]: `lhs: Non-...` according to coding style

File: src/goto-symex/symex_assign.cpp
Comment (peterschrammel) [inline src/goto-symex/symex_assign.cpp]: `\return The...` according to coding style

File: src/goto-symex/symex_target.h
Comment (majakusber) [inline src/goto-symex/symex_target.h]: :pick: Typo - `similar to goto_programt::const_target which is ..`?

File: src/goto-symex/symex_target.h
Comment (peterschrammel) [inline src/goto-symex/symex_target.h]: It would increase readability to indent here.

File: src/goto-symex/symex_target.h
Comment (majakusber) [inline src/goto-symex/symex_target.h]: :pick: Consider using imperative mood in all function docs - here `Read` instead of `Reading`. I see that further below you do use it, so would be good to keep it consistent. :grapes:

File: src/goto-symex/symex_target.h
Comment (majakusber) [inline src/goto-symex/symex_target.h]: :pick: This sentence is strange, how about `precondition for this read event` :strawberry:

File: src/goto-symex/symex_target.h
Comment (majakusber) [inline src/goto-symex/symex_target.h]: :strawberry:

File: src/goto-symex/symex_target.h
Comment (majakusber) [inline src/goto-symex/symex_target.h]: :strawberry:

File: src/goto-symex/symex_target.h
Comment (majakusber) [inline src/goto-symex/symex_target.h]: :strawberry:

File: src/goto-symex/symex_target.h
Comment (majakusber) [inline src/goto-symex/symex_target.h]: :strawberry:

File: src/goto-symex/symex_target.h
Comment (majakusber) [inline src/goto-symex/symex_target.h]: :strawberry:

File: src/goto-symex/symex_target.h
Comment (majakusber) [inline src/goto-symex/symex_target.h]: :strawberry:

File: src/goto-symex/symex_target.h
Comment (majakusber) [inline src/goto-symex/symex_target.h]: :strawberry:

File: src/goto-symex/symex_target.h
Comment (majakusber) [inline src/goto-symex/symex_target.h]: :strawberry:

File: src/goto-symex/symex_target.h
Comment (majakusber) [inline src/goto-symex/symex_target.h]: :strawberry:

File: src/goto-symex/symex_target.h
Comment (majakusber) [inline src/goto-symex/symex_target.h]: :strawberry:

File: src/goto-symex/symex_target.h
Comment (majakusber) [inline src/goto-symex/symex_target.h]: :strawberry:

File: src/goto-symex/symex_target.h
Comment (majakusber) [inline src/goto-symex/symex_target.h]: :strawberry:

File: src/goto-symex/symex_target.h
Comment (majakusber) [inline src/goto-symex/symex_target.h]: :strawberry:

File: src/goto-symex/symex_target.h
Comment (majakusber) [inline src/goto-symex/symex_target.h]: :question: Why are the last four functions not documented?

File: src/goto-symex/symex_assign.cpp
Comment (majakusber) [inline src/goto-symex/symex_assign.cpp]: :green_apple:

File: src/goto-symex/symex_target_equation.cpp
Comment (majakusber) [inline src/goto-symex/symex_target_equation.cpp]: :pick: Indent by 2 spaces for better readability. :banana:

File: src/goto-symex/symex_target_equation.cpp
Comment (majakusber) [inline src/goto-symex/symex_target_equation.cpp]: :grapes:

File: src/goto-symex/symex_target_equation.h
Comment (majakusber) [inline src/goto-symex/symex_target_equation.h]: :pick: `are only assigned once` (without the `to`)

Comment (majakusber) [reply src/goto-symex/symex_target_equation.h]: :bulb: Might be useful to link to `\ref static-single-assignment` here (SSA documentation in the `goto_symex` folder readme)

File: src/goto-symex/symex_target_equation.h
Comment (majakusber) [inline src/goto-symex/symex_target_equation.h]: :pick: Represents the SSA form of what? How does it extend the base class?

**Review by majakusber — COMMENTED** [review]
Generally looks good, a lot of useful docs added! I have mostly minor suggestions for formatting (sorry for all the fruits :slightly_smiling_face: first I wanted to mark all the places for the recurring comments with them but it's too much), here is a summary:
- use imperative mood in function docs,
- indent the overflowing lines by two spaces for better readability,
- move docs of interface functions to headers,
- I would suggest calling the `guard`s preconditions and reformulating the sentence a bit (see :strawberry:).
There are few other comments but nothing too big.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: c0b43dae325d222ccb1fb226979582b7a64d48b2).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/91244044

File: src/goto-symex/symex_assign.cpp
Comment (tautschnig) [reply src/goto-symex/symex_assign.cpp]: How about using `\p` to also make clear this refers to a parameter (and not just some arbitrary code)?

File: src/goto-symex/symex_target.h
Comment (tautschnig) [inline src/goto-symex/symex_target.h]: What is a "position identifier?"

File: src/goto-symex/symex_assign.cpp
Comment (xbauch) [reply src/goto-symex/symex_assign.cpp]: Done (throughout the PR).

File: src/goto-symex/symex_target.cpp
Comment (xbauch) [reply src/goto-symex/symex_target.cpp]: Done.

File: src/goto-symex/symex_target.h
Comment (xbauch) [reply src/goto-symex/symex_target.h]: Done.

File: src/goto-symex/symex_target.h
Comment (xbauch) [reply src/goto-symex/symex_target.h]: Done.

File: src/goto-symex/symex_target.h
Comment (xbauch) [reply src/goto-symex/symex_target.h]: Done.

File: src/goto-symex/symex_target.h
Comment (xbauch) [reply src/goto-symex/symex_target.h]: Hopefully all the strawberries should be handled.

File: src/goto-symex/symex_assign.cpp
Comment (xbauch) [reply src/goto-symex/symex_assign.cpp]: And apples as well.

File: src/goto-symex/symex_target_equation.cpp
Comment (xbauch) [reply src/goto-symex/symex_target_equation.cpp]: Done, and hopefully all the bananas.

File: src/goto-symex/symex_target_equation.h
Comment (xbauch) [reply src/goto-symex/symex_target_equation.h]: Done.

File: src/goto-symex/symex_target_equation.h
Comment (xbauch) [reply src/goto-symex/symex_target_equation.h]: Done.

File: src/goto-symex/symex_assign.cpp
Comment (xbauch) [reply src/goto-symex/symex_assign.cpp]: A swapped backslash-c with back-ticks throughout the PR.

File: src/goto-symex/symex_assign.cpp
Comment (xbauch) [reply src/goto-symex/symex_assign.cpp]: Done, here and elsewhere.

File: src/goto-symex/symex_assign.cpp
Comment (xbauch) [reply src/goto-symex/symex_assign.cpp]: Also done.

File: src/goto-symex/symex_target.h
Comment (xbauch) [reply src/goto-symex/symex_target.h]: Done.

File: src/goto-symex/symex_target.h
Comment (xbauch) [reply src/goto-symex/symex_target.h]: I wasn't sure if we have an established term to refer to program location. To the best of my understanding `symex_targett` has a similar purpose to `const_target` which is an iterator into a list of instructions. I tried rewording the description a bit.

Comment (tautschnig) [reply src/goto-symex/symex_target.h]: Oh, I see, actually a `symex_targett` is very different (although by now I'm not even sure it is a meaningful interface anymore, because we leak out `symex_target_equationt` all over the place). A `symex_targett` is the interface of the target _container_ for symbolic execution to record its symbolic steps into. As far as I am aware, `symex_target_equationt` is the only implementation of that interface.

File: src/goto-symex/symex_target_equation.cpp
Comment (majakusber) [reply src/goto-symex/symex_target_equation.cpp]: Thanks, I think only one space is added now but that's probably enough :slightly_smiling_face: (maybe you ran into the same mistake I sometimes make - hitting tab right after the `///` results in adding one space only but hitting tab at the place where text starts one space after the `///` indents by 2 spaces then)

File: src/goto-symex/symex_target.h
Comment (majakusber) [reply src/goto-symex/symex_target.h]: This is still outstanding

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: c3b447b23808b00113517d4ecc06ca324e258c01).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/91500113

File: src/goto-symex/symex_target.h
Comment (xbauch) [reply src/goto-symex/symex_target.h]: I hope you won't mind if I quote/paraphrase the above for the description.

**Comment by xbauch** [comment]
I think I've addressed all the @svorenova @peterschrammel and @tautschnig comments.

File: src/goto-symex/symex_target.h
Comment (majakusber) [inline src/goto-symex/symex_target.h:249]: :question: I asked this before but it may have gotten lost - why are the last four functions below not documented?

**Review by majakusber — APPROVED** [review]
Looks good to go, I just have one question about four functions without doc. But because this is already much better than what it was before I won't block the merge.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 7d771982cea6865eff6416e5bb22ca7baedbb688).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/91549113

File: src/goto-symex/symex_target.h
Comment (peterschrammel) [reply src/goto-symex/symex_target.h]: > `symex_target_equationt` is the only implementation of that interface.

Yes, afaik.

**Comment by peterschrammel** [comment]
@xbauch, can you squash the commits please?

File: src/goto-symex/symex_target.h
Comment (xbauch) [reply src/goto-symex/symex_target.h:249]: Added.

**Comment by xbauch** [comment]
Squashed.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: fffc954d5b5ad1b67dafe1879a889aff0adc2c0e).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/91638896

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: e399467856dd196fb9c7adb19b95495ba5423645).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/93174124
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Comment by tautschnig** [comment]
@xbauch Could you please rebase to make sure the bot is re-triggered as the past failure was likely caused by unrelated problems?

File: src/goto-symex/goto_symex.h
Comment (tautschnig) [inline src/goto-symex/goto_symex.h]:
```suggestion
  /// \param what: The expression to be added to the \p lhs
```

File: src/goto-symex/goto_symex.h
Comment (tautschnig) [inline src/goto-symex/goto_symex.h]: What is an "unassigned" operand?

File: src/goto-symex/symex_target.h
Comment (tautschnig) [inline src/goto-symex/symex_target.h]:
```suggestion
/// symbolic steps into. Presently, \ref symex_target_equationt is the only
```

File: src/goto-symex/symex_target.h
Comment (tautschnig) [inline src/goto-symex/symex_target.h:32]: We might have to rethink the naming here - this really is a generalisation of the program counter with some additional context. But that's not for this PR.

File: src/goto-symex/symex_target.h
Comment (tautschnig) [inline src/goto-symex/symex_target.h]:
```suggestion
  /// Read from a shared variable \p ssa_rhs (which will thus become a
```

File: src/goto-symex/symex_target.h
Comment (tautschnig) [inline src/goto-symex/symex_target.h]:
```suggestion
  /// the object underlying \p ssa_rhs by another thread to \p ssa_rhs in the memory scope of this
```

File: src/goto-symex/symex_target.h
Comment (tautschnig) [inline src/goto-symex/symex_target.h]:
```suggestion
  ///  takes place (if any)
```

File: src/goto-symex/symex_target.h
Comment (tautschnig) [inline src/goto-symex/symex_target.h]: As is done in the implementation, please rename "ssa_rhs" to "ssa_object" and then remove the parts in parentheses - this always was a left-hand side, it's not actually becoming one here.

File: src/goto-symex/symex_target.h
Comment (tautschnig) [inline src/goto-symex/symex_target.h]:
```suggestion
  ///  takes place (if any)
```

File: src/goto-symex/symex_target_equation.h
Comment (tautschnig) [inline src/goto-symex/symex_target_equation.h]:
```suggestion
/// form of the input program as a list of \ref SSA_stept. It further extends the
```

File: src/goto-symex/symex_target_equation.h
Comment (tautschnig) [inline src/goto-symex/symex_target_equation.h]: Here and below: use `\copydoc`

**Review by tautschnig — APPROVED** [review]
A couple of suggestions and requests for changes below, but nothing blocking. The biggest item is using `\copydoc` to avoid duplicating the same text when functions are just implementations of an interface. In those cases please make sure the documentation genuinely was the same, else merge any missing bits into the documentation of the interface.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: bff5bf4819b8533633a9a204fe94f24f767ab50c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/93351125

**Pull request merged**
